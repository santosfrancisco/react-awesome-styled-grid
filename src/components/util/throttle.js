/**
const handleKeydown = throttle((e) => {
  console.log(e.target.value)
}, 300)
input.addEventListener('keydown', handleKeydown)
*/
export default function throttle (callback, wait, context = this) {
  let timeout = null
  let callbackArgs = null

  const later = () => {
    callback.apply(context, callbackArgs)
    timeout = null
  }

  return function () {
    if (!timeout) {
      callbackArgs = arguments
      timeout = setTimeout(later, wait)
    }
  }
}

/**
 * Normal event
 * event      | |      |
 * time     ----------------
 * callback   | |      |
 *
 * Call search at most once per 300ms while keydown
 * keydown     | |     |  |
 * time     -----------------
 * search          |       |
 *             |300|   |300|
 */
