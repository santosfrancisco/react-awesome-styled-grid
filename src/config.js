import { css } from 'styled-components'

const CUSTOM_CONF = 'awesomegrid'
export const DIMENSIONS = ['xs', 'sm', 'md', 'lg', 'xl']
export const BASE_CONF = {
  mediaQuery: 'only screen',
  columns: {
    xs: 4,
    sm: 8,
    md: 8,
    lg: 12,
    xl: 12
  },
  gutterWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  paddingWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  container: {
    xs: 'full', // 'full' = max-width: 100%
    sm: 'full', // 'full' = max-width: 100%
    md: 'full', // 'full' = max-width: 100%
    lg: 90, // max-width: 1440px
    xl: 90 // max-width: 1440px
  },
  breakpoints: {
    xs: 1,
    sm: 48, // 768px
    md: 64, // 1024px
    lg: 90, // 1440px
    xl: 120 // 1920px
  }
}

const configs = []
const hasCustomConf = props => JSON.stringify((props.theme && props.theme[CUSTOM_CONF]) || {})
const resolveConfig = props => {
  const themeConf = (props.theme && props.theme[CUSTOM_CONF]) || {}

  const conf = {
    ...BASE_CONF,
    ...themeConf
  }

  conf.media = Object.keys(conf.breakpoints).reduce((media, breakpoint) => {
    const breakpointWidth = conf.breakpoints[breakpoint]
    media[breakpoint] = makeMedia(
      [
        conf.mediaQuery,
        breakpointWidth >= 0 && `(min-width: ${breakpointWidth}rem)`
      ]
        .filter(Boolean)
        .join(' and ')
    )
    return media
  }, {})

  return conf
}

export default function config (props = {}) {
  const customConf = hasCustomConf(props)
  if (configs[0] === customConf) {
    return configs[1]
  }

  const conf = resolveConfig(props)

  configs[0] = customConf
  configs[1] = conf

  return conf
}

function makeMedia (media) {
  return (...args) => css`
    @media ${media} {
      ${css(...args)}
    }
  `
}
