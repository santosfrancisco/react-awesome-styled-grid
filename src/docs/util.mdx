---
name: Utilities
route: /util
order: 2
---

# Utilities

## Media Queries

To use in your css the media queries with the same breakpoints of the grid,
just import the config as in the example below. If you are not using a
custom config, it is not mandatory to pass *props* to config

```js
import React from 'react'
import styled from 'styled-components'
import {Container, Row, Col, config} from 'react-awesome-styled-grid'

export default styled(({ className, children}) => {
  return (
    <Container className={className}>
      <Row>
        <Col>{children}</Col>
      </Row>
    </Container>
  )
})`
  // css here...

  ${props => config(props).media['sm']`
    color: red;
    // ...more css here
  `}

  ${props => config(props).media['md']`
    color: purple;
    font-size: 1.5rem;
    // ...more css here
  `}
`
```

> **Your code**
>
>```js
>${props => config(props).media['sm']`
>  color: red;
>  // ...more css here
>`}
>```
>
>**Your output**
>
>```css
>@media only screen and (min-width: 48rem) {
>  color: red;
>  // ...more css here
>}
>```
