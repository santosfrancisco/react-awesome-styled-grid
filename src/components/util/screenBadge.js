import styled, {
  css
} from 'styled-components'

import config, {
  DIMENSIONS
} from '../../config'

const ScreenBadge = styled.div`
    z-index: 10;
    position: fixed;
    font-size: 1.5rem;
    font-weight: bold;
    right: 10px;
    bottom: 10px;
    width: 50px;
    height: 30px;
    background-color: #5901ad40;
    display: flex;
    border-radius: 5px;
    justify-content: center;
    align-items: center;
  ${p => css`
    ${DIMENSIONS.map(d =>
    config(p).breakpoints[d] && config(p).media[d]`
      ::before {
        content: '${d}'
      }
    `)}
  `}
`

ScreenBadge.displayName = 'ScreenBadge'

export default ScreenBadge
