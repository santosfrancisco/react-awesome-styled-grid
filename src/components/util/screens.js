import { DIMENSIONS, BREAKPOINTS } from '../../config'

export function getViewPort () {
  if (typeof window !== 'undefined' && window.innerWidth) {
    return window.innerWidth
  }
  return null
};

const theBreakpoints = BREAKPOINTS
const theScreenClasses = DIMENSIONS

export function getScreenClass () {
  let screenClass
  const viewport = getViewPort()
  if (viewport) {
    screenClass = theScreenClasses[0]
    if (theBreakpoints[1] && viewport >= theBreakpoints[1]) screenClass = theScreenClasses[1]
    if (theBreakpoints[2] && viewport >= theBreakpoints[2]) screenClass = theScreenClasses[2]
    if (theBreakpoints[3] && viewport >= theBreakpoints[3]) screenClass = theScreenClasses[3]
    if (theBreakpoints[4] && viewport >= theBreakpoints[4]) screenClass = theScreenClasses[4]
  }
  return screenClass
};
