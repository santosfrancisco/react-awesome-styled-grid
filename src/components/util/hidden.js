import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { getScreenClass } from '../util/screens'
import { withTheme } from 'styled-components'

export class Hidden extends Component {
  static defaultProps = {
    xs: false,
    sm: false,
    md: false,
    lg: false,
    xl: false,
    theme: {}
  }
  constructor () {
    super()
    this.state = {screen: 'xs'}
  }
  componentWillMount () {
    this.setScreen()
    if (typeof window !== 'undefined') {
      window.addEventListener('orientationchange', this.setScreen, false)
      window.addEventListener('resize', this.setScreen, false)
    }
  }
  setScreen = () => {
    let lastScreenClass = this.state.screen
    const actualScreenClass = getScreenClass(this.props)
    if (!lastScreenClass || (lastScreenClass !== actualScreenClass)) {
      lastScreenClass = actualScreenClass
      this.setState({screen: actualScreenClass})
    }
  }

  componentWillUnmount () {
    window.removeEventListener('orientationchange', this.setScreen)
    window.removeEventListener('resize', this.setScreen)
}

  isHidden () {
    const { xs, sm, md, lg, xl } = this.props
    switch (this.state.screen) {
      case 'xs':
        return xs
      case 'sm':
        return sm
      case 'md':
        return md
      case 'lg':
        return lg
      case 'xl':
        return xl
    }
  }

  render () {
    const { children } = this.props

    if (this.isHidden()) return false
    return (
      <React.Fragment>
        {children}
      </React.Fragment>
    )
  }
}

Hidden.displayName = 'Hidden'

Hidden.propTypes = {
  xs: PropTypes.bool,
  sm: PropTypes.bool,
  md: PropTypes.bool,
  lg: PropTypes.bool,
  xl: PropTypes.bool,
  children: PropTypes.node
}

export default withTheme(Hidden)
