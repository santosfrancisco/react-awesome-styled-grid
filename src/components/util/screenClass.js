import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { getScreenClass } from '../util/screens'
import { withTheme } from 'styled-components'

export class ScreenClass extends Component {
  static defaultProps = {
    theme: {}
  }
  constructor () {
    super()
    this.state = {screen: 'xs'}
  }
  componentWillMount () {
    this.setScreen()
    if (typeof window !== 'undefined') {
      window.addEventListener('orientationchange', this.setScreen, false)
      window.addEventListener('resize', this.setScreen, false)
    }
  }
  setScreen = () => {
    let lastScreenClass = this.state.screen
    const actualScreenClass = getScreenClass(this.props)
    if (!lastScreenClass || (lastScreenClass !== actualScreenClass)) {
      lastScreenClass = actualScreenClass
      this.setState({screen: actualScreenClass})
    }
  }

  componentWillUnmount () {
    window.removeEventListener('orientationchange', this.setScreen)
    window.removeEventListener('resize', this.setScreen)
  }

  render () {
    const { render } = this.props

    return (
      <React.Fragment>
        {render && render(this.state.screen)}
      </React.Fragment>
    )
  }
}

ScreenClass.displayName = 'ScreenClass'

ScreenClass.propTypes = {
  render: PropTypes.func
}

export default withTheme(ScreenClass)
