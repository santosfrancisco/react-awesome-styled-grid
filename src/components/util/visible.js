import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { getScreenClass } from '../util/screens'

class Visible extends Component {
  constructor () {
    super()
    this.state = {screen: 'xs'}
    this.setScreen = this.setScreen.bind(this)
  }
  componentDidMount () {
    this.setScreen()
    if (typeof window !== 'undefined') {
      window.addEventListener('orientationchange', this.setScreen, false)
      window.addEventListener('resize', this.setScreen, false)
    }
  }
  setScreen () {
    let lastScreenClass = this.state.screen
    const actualScreenClass = getScreenClass()
    if (!lastScreenClass || (lastScreenClass !== actualScreenClass)) {
      lastScreenClass = actualScreenClass
      this.setState({screen: actualScreenClass})
    }
  }

  componentWillUnmount () {
    window.removeEventListener('orientationchange', this.setScreen)
    window.removeEventListener('resize', this.setScreen)
  }

  isVisible () {
    const { xs, sm, md, lg, xl } = this.props
    switch (this.state.screen) {
      case 'xs':
        return xs
      case 'sm':
        return sm
      case 'md':
        return md
      case 'lg':
        return lg
      case 'xl':
        return xl
      default:
        return false
    }
  }

  render () {
    const { children } = this.props
    return (
      <React.Fragment>
        {this.isVisible() && children}
      </React.Fragment>
    )
  }
}

Visible.displayName = 'Visible'

Visible.propTypes = {
  xs: PropTypes.bool,
  sm: PropTypes.bool,
  md: PropTypes.bool,
  lg: PropTypes.bool,
  xl: PropTypes.bool,
  children: PropTypes.node
}

export default Visible
