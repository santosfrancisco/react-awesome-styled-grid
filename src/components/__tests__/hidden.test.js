import React from 'react'
import renderer from 'react-test-renderer'
import config from '../../config'
import {Hidden} from '../util/hidden'

describe('<Hidden />', () => {
  const resize = width => {
    global.window.innerWidth = width
    global.window.dispatchEvent(new Event('resize'))
  }

  it('should render default style correctly', () => {
    const tree = renderer.create(<Hidden />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should initialize state', () => {
    resize(1)
    const root = renderer.create(<Hidden xs>hidden</Hidden>).root
    expect(root.children).toHaveLength(0)
    resize(768)
    expect(root.children).toContain('hidden')
  })

  it('should unlisten resize and orientation events', () => {
    const spy = jest.spyOn(global.window, 'removeEventListener')

    const component = renderer.create(<Hidden />)
    const instance = component.getInstance()
    component.unmount()

    expect(spy).toHaveBeenCalledWith('orientationchange', instance.setScreen)
    expect(spy).toHaveBeenCalledWith('resize', instance.setScreen)

    spy.mockRestore()
  })

  it('should control screen state', () => {
    resize(1)
    const root = renderer.create(<Hidden sm>hidden</Hidden>).root
    expect(root.children).toContain('hidden')
    
    resize(768)
    expect(root.children).toHaveLength(0)
  })

  it('should always be hidden', () => {
    const root = renderer.create(<Hidden xs sm md lg xl >hidden</Hidden>).root
    Object.values(config().breakpoints).forEach(b => {
      resize(b * 16)
      expect(root.children).toHaveLength(0)
    })
  })
})
