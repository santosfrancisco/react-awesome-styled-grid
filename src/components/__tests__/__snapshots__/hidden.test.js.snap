// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hidden /> should render default style correctly 1`] = `null`;
